src/Microsoft.Health.Fhir.SqlServer/Features/Operations/Import/SqlImporter.cs

```diff
@@ -69,45 +73,158 @@ await foreach (ImportResource resource in inputChannel.Reader.ReadAllAsync(cance
+                        }
 
-            errors.AddRange(resourcesWithError.Select(r => r.ImportError));
-            AppendDuplicateErrorsToBuffer(dupsNotMerged, errors);
+                        Task.Delay(5000, cancellationToken);
+                        continue;
+                    }
 
-            succeedCount += mergedResources.Count();
-            failedCount += resourcesWithError.Count() + dupsNotMerged.Count();
+                    _logger.LogError(e, $"Error on {nameof(ImportResourcesInBufferMain)} retries={{Retries}}", retries);
+                    if (sqlEx != null)
+                    {
+                        _store.TryLogEvent(nameof(ImportResourcesInBufferMain), "Error", $"retries={retries} error={e}", mergeStart, cancellationToken).Wait();
+                    }
+
+                    throw;
+                }
+            }
+
+            var errorResources = resources.Where(r => !string.IsNullOrEmpty(r.ImportError));
+            errors.AddRange(errorResources.Select(r => r.ImportError));
+            var dups = resources.Where(r => string.IsNullOrEmpty(r.ImportError)).Except(loaded).Except(conflicts);
+            AppendErrorsToBuffer(dups, conflicts, errors);
+
+            succeededCount += loaded.Count;
+            failedCount += errorResources.Count() + dups.Count() + conflicts.Count;
+            processedBytes += resources.Sum(_ => (long)_.Length);
 
             resources.Clear();
         }
 
-        private void AppendDuplicateErrorsToBuffer(IEnumerable<ImportResource> resources, List<string> importErrorBuffer)
+        private async Task ImportResourcesInBufferMain(List<ImportResource> resources, List<ImportResource> loaded, List<ImportResource> conflicts, ImportMode importMode, CancellationToken cancellationToken)
+        {
+            var goodResources = resources.Where(r => string.IsNullOrEmpty(r.ImportError)).ToList();
+            if (importMode == ImportMode.InitialLoad)
+            {
+                var inputDedupped = goodResources.GroupBy(_ => _.ResourceWrapper.ToResourceKey(true)).Select(_ => _.OrderBy(_ => _.ResourceWrapper.LastModified).First()).ToList();
+                var current = new HashSet<ResourceKey>((await _store.GetAsync(inputDedupped.Select(_ => _.ResourceWrapper.ToResourceKey(true)).ToList(), cancellationToken)).Select(_ => _.ToResourceKey(true)));
+                loaded.AddRange(inputDedupped.Where(i => !current.TryGetValue(i.ResourceWrapper.ToResourceKey(true), out _)));
+                await MergeResourcesAsync(loaded, cancellationToken);
+            }
+            else
+            {
+                // dedup by last updated
+                var inputDedupped = goodResources.GroupBy(_ => _.ResourceWrapper.ToResourceDateKey(true)).Select(_ => _.First()).ToList();
+
+                // 2 paths:
+                // 1 - if versions were specified on input then dups need to be checked within input and database
```

Comment:
Do we have a check anywhere that the versions being provided are greater than 0?